pkg/controllers/clusters/clusterlifecycle/clusterlifecycle_controller.go

```diff
@@ -0,0 +1,360 @@
+		// do not want this work item being re-queued. For example, we do
+		// not call Forget if a transient error occurs, instead the item is
+		// put back on the workqueue and attempted again after a back-off
+		// period.
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		// We expect strings to come off the workqueue. These are of the
+		// form namespace/name. We do this as the delayed nature of the
+		// workqueue means the items in the informer cache may actually be
+		// more up to date that when the item was initially put onto the
+		// workqueue.
+		if key, ok = obj.(string); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// ManagedCluster resource to be synced.
+		requeueAfter, err := c.syncHandler(key)
+		switch {
+		case err != nil:
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		case requeueAfter > 0:
+			// Put the item back on the workqueue with delay.
+			c.workqueue.Forget(obj)
+			c.workqueue.AddAfter(obj, requeueAfter)
+		default:
+			c.workqueue.Forget(obj)
+		}
+		klog.Infof("successfully synced ManagedCluster %q", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the ManagedCluster resource
+// with the current status of the resource.
+// Duration <=0 means don't need to requeue this obj.
+func (c *Controller) syncHandler(key string) (time.Duration, error) {
+	// If an error occurs during handling, we'll requeue the item so we can
+	// attempt processing again later. This could have been caused by a
+	// temporary network failure, or any other transient reason.
+
+	// Convert the namespace/name string into a distinct namespace and name
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return time.Duration(0), nil
+	}
+
+	klog.V(4).Infof("start processing ManagedCluster %q", key)
+	// Get the ManagedCluster resource with this name
+	mcls, err := c.clusterLister.ManagedClusters(ns).Get(name)
+	// The ManagedCluster resource may no longer exist, in which case we stop processing.
+	if apierrors.IsNotFound(err) {
+		klog.V(2).Infof("ManagedCluster %q has been deleted", key)
+		return time.Duration(0), nil
+	}
+	if err != nil {
+		return time.Duration(0), err
+	}
+
+	err = c.updateClusterReadyConditionIfNeed(c.ctx, mcls, retry.DefaultBackoff)
```

Comment:
So this will always update cluster condition from LOST to READY when the hub gets a restart? Even if the cluster had stopped reporting heartbeats for quite a long long while.

Reply:
Only update to UNKNOWN if cluster gets lost

Reply:
Change the function name to avoid confusion:  updateClusterToLostIfNeeded